security/pkg/nodeagent/cache/secretcache.go

```diff
@@ -0,0 +1,195 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package cache is the in-memory secret store.
+package cache
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"istio.io/istio/pkg/log"
+	"istio.io/istio/security/pkg/pki/util"
+)
+
+// The size of a private key for a leaf certificate.
+const keySize = 2048
+
+// CAClient interface defines the clients need to implement to talk to CA for CSR.
+// TODO(quanlin): CAClient here is a placeholder, will move it to separated pkg.
+type CAClient interface {
+	Sign(csrPEM []byte /*PEM-encoded certificate request*/, subjectID string, certValidTTLInSec int64) ([]byte /*PEM-encoded certificate chain*/, error)
+}
+
+// SecretItem is the cached item in in-memory secret store.
+type SecretItem struct {
+	CertificateChain []byte
+	PrivateKey       []byte
+	token            string
+	lastUsedTime     time.Time
+	createdTime      time.Time
+}
+
+// SecretCache is the in-memory cache for secrets.
+type SecretCache struct {
+	// secrets map is the cache for secrets.
+	// map key is Envoy instance ID, map value is secretItem.
+	secrets        sync.Map
+	rotationTicker *time.Ticker
+	caClient       CAClient
+	closing        chan bool
+
+	// Cached secret will be removed from cache if (time.now - secretItem.lastUsedTime >= evictionDuration), this prevents cache growing indefinitely.
+	evictionDuration time.Duration
+
+	// Key rotation job running interval.
+	rotationInterval time.Duration
+
+	// secret TTL.
+	secretTTL time.Duration
+
+	// How may times that key rotation job has detected secret change happened, used in unit test.
+	secretChangedCount uint64
+}
+
+// NewSecretCache creates a new secret cache.
+func NewSecretCache(cl CAClient, secretTTL, rotationInterval, evictionDuration time.Duration) *SecretCache {
+	ret := &SecretCache{
+		caClient:         cl,
+		rotationInterval: rotationInterval,
+		evictionDuration: evictionDuration,
+		secretTTL:        secretTTL,
+		closing:          make(chan bool, 1),
+	}
+
+	atomic.StoreUint64(&ret.secretChangedCount, 0)
+	go ret.keyCertRotationJob()
+	return ret
+}
+
+// GetSecret gets secret from cache.
+func (sc *SecretCache) GetSecret(proxyID, token string) (*SecretItem, error) {
```

Comment:
Do you want  to use proxyID(assuming it means the envoy instance) as the key? How is that information obtained by node agent?  Can we use the identity itself as the key? Which seems a bit more straightforward to me, since the key/certs are bound with an identity, more than an instance.

Reply:
discovery request passing to SDS contains proxyID, I think it's OK for it in this initial PR and we could tweak it later once having better integration picture.